README.md
# Encodex
![example workflow](https://github.com/grqphical07/encodex/actions/workflows/rust.yml/badge.svg)

A simple CLI tool to encode text

## Installation

Clone this repo then run
```bash
$ cargo install --path .
```

## Usage

To encode text into Base64 run:
```bash
$ encodex "Hello World" -e base64
```

It should print
```SGVsbG8gV29ybGQ=```

To decode values just add the flag ```-d```
```bash
$ encodex "SGVsbG8gV29ybGQ=" -e base64 -d
```

It should print
```Hello World```

## Encoding formats

Currently encodex supports 4 formats

| **Format**     | **Name**   | **Example (Encoded the phrase "Hello World")**                                  |
|-------------|--------|----------------------------------------------------------------------------------------|
| Base64      | base64 | SGVsbG8gV29ybGQ=                                                                       |
| Binary      | binary | 1001000 1100101 1101100 1101100 1101111 100000 1010111 1101111 1110010 1101100 1100100 |
| Bytes       | bytes  | 72 101 108 108 111 32 87 111 114 108 100                                               |
| Hexadecimal | hex    | 48656c6c6f20576f726c64                                                                 |

## Changelog

[Changelog](https://github.com/grqphical07/encodex/blob/master/Changelog.md)


Changelog.md
## Changelog

## Version 0.1.1 - September 22, 2023

- Added formatted flag to toggle whether or not you want output to be formatted. Ex: `72 101 108 108 111 32 87 111 114 108 100` instead of `721011081081113287111114108100`
- This feature only works on the `binary` and `bytes` encoding

src/encodings/bytes.rs
use crate::{handle_error, EXIT_INVALID_DATA};

pub fn encode(data: &[u8], formatted: bool) -> String {
    let mut output = String::new();
    for byte in data {
        if formatted {
            output.push_str(&format!("{} ", byte))
        } else {
            output.push_str(&format!("{}", byte))
        }
        
    }

    output = output.trim_end().into();

    output
}

pub fn decode(data: String) -> String {
    let bytes: Vec<u8> = data.split(" ").map(|x| {
        let result = x.parse::<u8>();
        match result {
            Ok(value) => value,
            Err(_) => {
                handle_error("Invalid byte data".into(), EXIT_INVALID_DATA);
                0
            }
        }
    }).collect();
    let mut output = String::new();

    for byte in bytes {
        output.push(byte as char)
    }
    output
}

src/encodings/base64.rs
use crate::{handle_error, EXIT_INVALID_DATA};

const BASE_CHARS: &[u8] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

pub fn encode(input: &[u8]) -> String {
    // Implementation based on https://keyboardsmash.dev/posts/implementing-base64-algorithm-in-rust/

    let mut output = Vec::new();
    let input_length = input.len();

    for i in (0..input_length).step_by(3) {
        let a = input.get(i).unwrap();
        let b = input.get(i + 1).unwrap_or(&0);
        let c = input.get(i + 2).unwrap_or(&0);

        let enc1 = (a >> 2) as usize;
        let enc2 = (((a & 0x3) << 4) | (b >> 4)) as usize;
        let enc3 = (((b & 0xf) << 2) | (c >> 6)) as usize;
        let enc4 = (c & 0x3f) as usize;

        output.push(BASE_CHARS[enc1]);
        output.push(BASE_CHARS[enc2]);
       
        output.push(BASE_CHARS[enc3]);
        output.push(BASE_CHARS[enc4]);
    }

    let output_len = output.len();
    let padding_len = match input_length % 3 {
        1 => 2, // Add two paddings
        2 => 1, // Add one padding
        _ => 0, // No paddings needed
    };

    for i in 0..padding_len {
        output[output_len - 1 - i] = b'=';
    }

    String::from_utf8(output).unwrap()
}

pub fn decode(input: &[u8]) -> String {
    // Implementation based on https://keyboardsmash.dev/posts/base64-implementation-in-rust-decoding/

	let mut output: Vec<u8> = Vec::new();
	for chunk in input.chunks(4) {

        if chunk.len() != 4 {
            handle_error("Invalid data provided".into(), EXIT_INVALID_DATA)
        }

    	let a = decode_char(chunk[0]);
    	let b = decode_char(chunk[1]);
    	let c = decode_char(chunk[2]);
    	let d = decode_char(chunk[3]);

    	let dec1 = ((a << 2) | (b & 0x30) >> 4) as u8;
    	let dec2 = (((b & 0x0F) << 4) | (c & 0x3C) >> 2) as u8;
    	let dec3 = (((c & 0x03) << 6) | (d)) as u8;

    	output.push(dec1);
    	output.push(dec2);
    	output.push(dec3);
	}

	String::from_utf8(output).unwrap().replace("\0", "")
}


fn decode_char(input: u8) -> u8 {
	BASE_CHARS.iter().position(|&c| c == input).unwrap_or(0) as u8
}

src/encodings/binary.rs
use crate::{handle_error, EXIT_INVALID_DATA};

pub fn encode(data: &[u8], formatted: bool) -> String {
    let mut output = String::new();

    for byte in data {
        if formatted {
            output.push_str(&format!("{:b} ", byte))
        } else {
            output.push_str(&format!("{:b}", byte))
        }
        
    }

    output = output.trim_end().into();

    output
}

pub fn decode(data: String) -> String {
    let chunks = data.split(" ");
    let mut output = String::new();

    for chunk in chunks {
        // Convert the string into a u8
        let conversion_result = u8::from_str_radix(chunk, 2);
        let mut integer_value: u8 = 0;

        match conversion_result {
            Ok(val) => integer_value = val,
            Err(_) => handle_error("Invalid Binary".into(), EXIT_INVALID_DATA)
        }

        // Add it to the result string as a character
        let char = integer_value as char;

        output.push(char)
    }

    output
}

src/main.rs
mod encodings;

use clap::{Parser, ValueEnum};
use encodings::{base64, binary, bytes, hex};
use std::{path::PathBuf, str};

// Tests
#[cfg(test)]
mod tests {
    use crate::encodings;
    #[test]
    fn base64_encode_test() {
        let test_data = "Hello World";
        let expected = "SGVsbG8gV29ybGQ=";

        assert_eq!(encodings::base64::encode(test_data.as_bytes()), expected)
    }
    #[test]
    fn base64_decode_test() {
        let test_data = "SGVsbG8gV29ybGQ=";
        let expected = "Hello World";

        assert_eq!(encodings::base64::decode(test_data.as_bytes()), expected)
    }
    #[test]
    fn binary_encode_test() {
        let test_data = "Hello World";
        let expected = "1001000 1100101 1101100 1101100 1101111 100000 1010111 1101111 1110010 1101100 1100100";

        assert_eq!(encodings::binary::encode(test_data.as_bytes(), true), expected)
    }

    #[test]
    fn binary_decode_test() {
        let test_data = "1001000 1100101 1101100 1101100 1101111 100000 1010111 1101111 1110010 1101100 1100100";
        let expected = "Hello World";

        assert_eq!(encodings::binary::decode(test_data.into()), expected)
    }

    #[test]
    fn bytes_decode_test() {
        let test_data = "72 101 108 108 111 32 87 111 114 108 100";
        let expected = "Hello World";

        assert_eq!(encodings::bytes::decode(test_data.into()), expected)
    }

    #[test]
    fn bytes_encode_test() {
        let test_data = "Hello World";
        let expected = "72 101 108 108 111 32 87 111 114 108 100";

        assert_eq!(encodings::bytes::encode(test_data.as_bytes(), true), expected)
    }

    #[test]
    fn hex_decode_test() {
        let test_data = "48656c6c6f20576f726c64";
        let expected = "Hello World";

        let decoding_result = encodings::hex::decode(test_data).unwrap();
        let s = std::str::from_utf8(&decoding_result).unwrap();

        assert_eq!(s, expected)
    }

    #[test]
    fn hex_encode_test() {
        let test_data = "Hello World";
        let expected = "48656c6c6f20576f726c64";

        assert_eq!(encodings::hex::encode(test_data.as_bytes()), expected)
    }
}

// EXIT CODES
const EXIT_FILE_NOT_FOUND: i32 = 2;
const EXIT_INVALID_UTF8: i32 = 3;
const EXIT_NO_DATA_PROVIDED: i32 = 4;
const EXIT_INVALID_DATA: i32 = 5;

/// Function to gracefully handle errors by printing a message and exiting
///
/// # Arguments
///
/// * `err` - The error message
///
/// * `code` - The error code to exit with
///
/// # Examples
///
/// ```
/// handle_error("Invalid data".into(), 1);
/// ```
pub fn handle_error(err: String, code: i32) {
    println!("\x1b[1;31mERROR: {}\x1b[0;0m", err);
    std::process::exit(code);
}

/// All avalible encoding formats
#[derive(ValueEnum, Clone, Debug)]
enum Encodings {
    Base64,
    Hex,
    Binary,
    Bytes,
}

/// Structure of how command line args should be parsed
#[derive(Parser, Debug)]
#[command(version, author = "grqphical07", about = "A simple tool to encode strings in different formats", long_about = None)]
struct Cli {
    /// Optional text to encode
    text: Option<String>,

    /// Optional file to encode
    #[arg(short, long, value_name = "FILE")]
    file: Option<PathBuf>,

    /// Whether or not you wish to encode or decode
    #[arg(short, long)]
    decode: bool,

    /// What encoding format do you wish to use
    #[arg(short, long)]
    encoding: Encodings,

    /// Whether or not the value should be formatted
    #[arg(long)]
    formatted: bool,
}

fn main() {
    // Parse the cli args
    let cli = Cli::parse();

    let mut text = String::new();

    // Check if user wants to read from a file and if so read from it
    // If no file is specified use the text the user entered into the application
    match cli.file {
        Some(val) => {
            let file_result = std::fs::read_to_string(val);
            match file_result {
                Ok(value) => text = value,
                Err(err) => handle_error(err.to_string(), EXIT_FILE_NOT_FOUND),
            }
        }
        None => match cli.text {
            Some(value) => text = value,
            None => handle_error(
                "No data provided. Use --help for options".into(),
                EXIT_NO_DATA_PROVIDED,
            ),
        },
    }

    // Encode/Decode the data based on which format the user wants to use
    match cli.encoding {
        Encodings::Base64 => {
            if cli.decode {
                println!("{}", base64::decode(text.as_bytes()))
            } else {
                println!("{}", base64::encode(text.as_bytes()))
            }
        }
        Encodings::Hex => {
            if cli.decode {
                let decoding_result = hex::decode(&text);
                let mut decoded_bytes: Vec<u8> = Vec::new();

                match decoding_result {
                    Ok(value) => decoded_bytes = value,
                    Err(_) => handle_error("Invalid Hex Data".into(), EXIT_INVALID_DATA),
                }

                let s = match str::from_utf8(&decoded_bytes) {
                    Ok(v) => v,
                    Err(_) => {
                        handle_error("Invalid Text Sequence".into(), EXIT_INVALID_UTF8);
                        ""
                    }
                };
                println!("{}", s)
            } else {
                println!("{}", hex::encode(text.as_bytes()))
            }
        }
        Encodings::Bytes => {
            if cli.decode {
                println!("{}", bytes::decode(text))
            } else {
                println!("{}", bytes::encode(text.as_bytes(), cli.formatted))
            }
        }
        Encodings::Binary => {
            if cli.decode {
                println!("{}", binary::decode(text))
            } else {
                println!("{}", binary::encode(text.as_bytes(), cli.formatted))
            }
        }
    }
}
